### ISSUE TITLE
Question about handling pretrained embeddings variables

### ISSUE THREAD

**Issue by mattivi** [shap/shap#440]
Hi, 
I am using SHAP to explain the output of an autoencoder model built with Keras.
As input I have many numerical and categorical variables.  
I am one-hot encoding variables with low cardinality, and substituting high cardinality variables with  embeddings learned on another task. E.g. instead of one-hot encoding or label encoding a variable with 8000 categories, I replace it with other 100 learned embedding features, as in this example:
- original: [numerical_1, categorical_1, ...]
- replaced with pretrained embeddings: [numerical_1, categorical_1_dim1, categorical_1_dim2, ..., categorical_1_dim100]

Following #397, in case of **one-hot encoded variable**:
- to explain a single category, we have already a SHAP value for the [category_0, category_1, category_2 ……]
- to explain the original variable BEFORE one-hot encoding, we can take the SUM over all one-hot encoded variables (SUM over [category_0, category_1, category_2 ……])

In case of **pretrained embeddings** [categorical_1_dim1, categorical_1_dim2, ..., categorical_1_dim100].. 
- to explain the original variable BEFORE embeddings, we can similarly take the SUM over all dimensions (SUM over [categorical_1_dim1, categorical_1_dim2, ..., categorical_1_dim100])

Is there any way that SHAP can explain the original category for the embedded feature? Back to the example, SHAP to provide a value to one or more of the original 8000 categories that were replaced with the pretrained embeddings. 

Thanks a lot for your suggestions!

**Comment by slundberg** [comment]
I have some ideas on this but need a bit more time to formulate/share them, will post back here later...

**Comment by jerry-rubikloud** [comment]
I have similar interests into this discussion. Basically, I use some raw features (e.g. age, gender, num_purchases etc) and I generate user embedding out of it. How can we formulate SHAP to assign values to each the raw features for a given user vector? Thank you.

**Comment by mattivi** [comment]
Hi, 
I have started to work again on the problem. I changed architecture, from the one below (where 106 features are the results of 100 embeddings dim + 6 numerical features):
![image](https://user-images.githubusercontent.com/1651448/54679897-5a628500-4b00-11e9-9e55-e73c6407e371.png )
to add an embedding layer that will map first the categorical features into their corresponding embeddings, and then the other numerical features are concatenated:
![image](https://user-images.githubusercontent.com/1651448/54679964-88e06000-4b00-11e9-86fa-eedc3290d420.png)

In this manner, the input for SHAP is only 6 features (without embeddings values), and the categorical features could then be treated by SHAP as in the ["Census income classification with scikit-learn" example](https://github.com/slundberg/shap/blob/master/notebooks/kernel_explainer/Census%20income%20classification%20with%20scikit-learn.ipynb)  
@slundberg is this what you had in mind?
Many thanks!
